### PR TITLE
Update label errors notebooks to properly use test/train splits

### LINKED ISSUE #11: Inconsistency in Dataset Used for acc_original and acc_clean Evaluations in Notebook Lab - Label Errors.ipynb
Hello,

I was reviewing the notebook "Lab - Label Errors.ipynb" and noticed an inconsistency in the datasets used for evaluating `acc_original` and `acc_clean`.

In the notebook, the line `data = df.drop(issue_idx)` results in the removal of rows with index `issue_idx` from the dataset. This effectively reduces the size of the dataset used for subsequent steps. As a result, the `acc_original` is evaluated on the full dataset, whereas `acc_clean` is evaluated on a subset of the original dataset (excluding the rows with index `issue_idx`).

From my understanding, for a fair comparison between `acc_original` and `acc_clean`, we should train without the problematic `issue_idx` rows but still make predictions on the entire dataset. By excluding these rows for the `acc_clean` evaluation, it seems like we are leaving out some of the trickiest rows to predict, which might lead to misleading results.

To address this, I've come up with a fix proposal:

```python
# ADDED START
class CL_XGBClassifier(XGBClassifier):
    def __init__(self, idx, **kwargs):
        super().__init__(**kwargs)
        self.idx = idx
        
    def fit(self, X, y):
        X['target'] = y
        for idx in self.idx:
            if idx in X.index:
                X = X.drop(idx)
        y = X.pop('target')
        return super().fit(X, y)
# ADDED END

# Remove the label errors found by Confident Learning
data = df.copy()  # <- MODIFIED
clean_labels = data['noisy_letter_grade']
data = data.drop(['stud_ID', 'letter_grade', 'noisy_letter_grade'], axis=1)

# Train a more robust classifier with less erroneous data
model = CL_XGBClassifier(tree_method="hist", enable_categorical=True, idx=issue_idx)    # <- MODIFIED
clean_pred_probs = cross_val_predict(model, data, clean_labels, method='predict_proba')
clean_preds = np.argmax(clean_pred_probs, axis=1)

acc_clean = accuracy_score(clean_preds, clean_labels)
print(f"Accuracy with original data: {round(acc_original*100, 1)}%")
print(f"Accuracy with errors found by Confident Learning removed: {round(acc_clean*100, 1)}%")

# Compute reduction in error.
err = ((1-acc_original)-(1-acc_clean))/(1-acc_original)
print(f"Reduction in error: {round(err*100,1)}%")
```

This proposal ensures that the training process avoids rows with problematic labels, but the predictions still cover the entire dataset, allowing for a fair comparison between `acc_original` and `acc_clean`.

Thank you for considering this feedback. I'm looking forward to any thoughts or clarifications on this matter.

Best regards,

RomainBdt

### PR DESCRIPTION
**Pull request by cmauck10**
Fixes #11

### CODE REVIEW

File: label_errors/solutions/Solution - Label Errors.ipynb [hunk 1]
```diff
@@ -765,24 +779,24 @@
...
+      "Reduction in error: 34.7%\n"
      ]
     }
    ],
    "source": [
-    "# Remove the label errors found by Confident Learning\n",
-    "data = df.drop(issue_idx)\n",
+    "# Remove the label errors found by Confident Learning from the train set\n",
+    "data = df_train.drop(issue_idx)\n",
     "clean_labels = data['noisy_letter_grade']\n",
```
Comment (jwmueller) [inline label_errors/solutions/Solution - Label Errors.ipynb:791]:
```suggestion
    "filtered_labels = data['noisy_letter_grade']\n",
```

File: label_errors/solutions/Solution - Label Errors.ipynb [hunk 2]
```diff
@@ -765,24 +779,24 @@
...
     "data = data.drop(['stud_ID', 'letter_grade', 'noisy_letter_grade'], axis=1)\n",
     "\n",
     "# Train a more robust classifier with less erroneous data\n",
     "model = XGBClassifier(tree_method=\"hist\", enable_categorical=True)\n",
-    "clean_pred_probs = cross_val_predict(model, data, clean_labels, method='predict_proba')\n",
-    "clean_preds = np.argmax(clean_pred_probs, axis=1)\n",
-    "\n",
-    "acc_clean = accuracy_score(clean_preds, clean_labels)\n",
+    "model.fit(data, clean_labels)\n",
```
Comment (jwmueller) [inline label_errors/solutions/Solution - Label Errors.ipynb:796]:
```suggestion
    "model.fit(data, filtered_labels)\n",
```

File: label_errors/solutions/Solution - Label Errors.ipynb (hunk 1 above)
Comment (jwmueller) [reply label_errors/solutions/Solution - Label Errors.ipynb:791]: clean_labels may be confused w correct/non-noisy labels

**Review by jwmueller — APPROVED** [review]
thanks for fixing this! I just had one suggested change, feel free to merge it after without another review from me
